src/Microsoft.Extensions.DependencyInjection.Ordered/Ordered.cs

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.DependencyInjection.Ordered
+{
+    internal class Ordered<T>: IOrdered<T>, IDisposable
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly OrderedEnumerableServiceDescriptorContainer<T> _descriptorContainer;
+        private List<T> _values;
+        private List<IDisposable> _dispose;
+
+        public Ordered(IServiceProvider serviceProvider, OrderedEnumerableServiceDescriptorContainer<T> descriptorContainer)
+        {
+            _serviceProvider = serviceProvider;
+            _descriptorContainer = descriptorContainer;
+        }
+
+        private void EnsureValues()
+        {
+            lock (_descriptorContainer)
+            {
+                if (_values != null)
+                {
+                    return;
+                }
+
+                _values = new List<T>();
+                _dispose = new List<IDisposable>();
+                foreach (var descriptor in _descriptorContainer.ServiceDescriptor.Descriptors)
+                {
+                    var factoryServiceDescriptor = descriptor as FactoryServiceDescriptor;
```

Comment:
Might be ever so slightly nicer to just do a `is` check (and cast) here

```
if (descriptor is FactoryServiceDescriptor)
{
  ...
}
```

Fewer levels of conditional nesting.